base/linalg/generic.jl

```diff
@@ -1165,8 +1165,46 @@ function logdet(A::AbstractMatrix)
     return d + log(s)
 end
 
+"""
+    promote_leaf_eltypes(itr)
+
+For an (possibly nested) iterable object `itr`, promote the types of leaf
+elements.  Equivalent to `promote_type(typeof(leaf1), typeof(leaf2), ...)`.
+
+# Example
+
+```jldoctest
+julia> a = [[1,2, [3,4]], 5.0, [6im, [7.0, 8.0]]]
+3-element Array{Any,1}:
+  Any[1,2,[3,4]]
+ 5.0
+  Any[0+6im,[7.0,8.0]]
+
+julia> promote_leaf_eltypes(a)
+Complex{Float64}
+```
+"""
+function promote_leaf_eltypes(itr)
+    if !isa(itr, AbstractArray) && !isa(itr, Tuple)  # handles arrays, tuples, ranges
+        return typeof(itr)
+    else
+        s = start(itr)
+        (i, s) = next(itr, s)
+        t = promote_leaf_eltypes(i)
+        while !done(itr, s)
```

Comment:
Thanks for the great suggestion!

One thing, though, is that I thought `promote_leaf_eltypes` could be useful in the context other than implementation of `isapprox`, so I wanted to make it work for any containers even with non-`Number` entries (e.g., so that we could get `String` for `promote_leaf_eltypes([["abc", "def"], "ghi"])`.  

Do you think the utility of `promote_leaf_eltypes` will be limited to implementation of `isapprox`?  If that's the case, I will change its implementation as you suggested.